File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-flutter/lib/src/mail_capability_catalog.dart
final class MailCapabilityCatalogEntry {
  const MailCapabilityCatalogEntry({
    required this.capabilityKey,
    required this.category,
    required this.surface,
  });

  final String capabilityKey;
  final String category;
  final String surface;
}

final class MailCapabilityCatalog {
  static const List<MailCapabilityCatalogEntry> entries = <MailCapabilityCatalogEntry>[
    MailCapabilityCatalogEntry(
      capabilityKey: "transport.connect",
      category: "required-baseline",
      surface: "control-plane",
    ),
    MailCapabilityCatalogEntry(
      capabilityKey: "transport.authenticate",
      category: "required-baseline",
      surface: "control-plane",
    ),
    MailCapabilityCatalogEntry(
      capabilityKey: "health",
      category: "required-baseline",
      surface: "control-plane",
    ),
    MailCapabilityCatalogEntry(
      capabilityKey: "smtp.send",
      category: "optional-advanced",
      surface: "runtime-bridge",
    ),
    MailCapabilityCatalogEntry(
      capabilityKey: "imap.sync",
      category: "optional-advanced",
      surface: "runtime-bridge",
    ),
    MailCapabilityCatalogEntry(
      capabilityKey: "imap.folder-sync",
      category: "optional-advanced",
      surface: "runtime-bridge",
    ),
    MailCapabilityCatalogEntry(
      capabilityKey: "imap.message-sync",
      category: "optional-advanced",
      surface: "runtime-bridge",
    ),
    MailCapabilityCatalogEntry(
      capabilityKey: "transport.retry",
      category: "optional-advanced",
      surface: "control-plane",
    ),
    MailCapabilityCatalogEntry(
      capabilityKey: "transport.pool",
      category: "optional-advanced",
      surface: "control-plane",
    )
  ];

  const MailCapabilityCatalog._();
}

List<MailCapabilityCatalogEntry> getMailCapabilityCatalog() {
  return MailCapabilityCatalog.entries;
}

MailCapabilityCatalogEntry? getMailCapabilityDescriptor(String capabilityKey) {
  for (final entry in MailCapabilityCatalog.entries) {
    if (entry.capabilityKey == capabilityKey) {
      return entry;
    }
  }

  return null;
}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-csharp/src/SDKWork.Mail.Sdk/mailCapabilityCatalog.cs
namespace Sdkwork.Mail.Sdk;

using System.Collections.Generic;
using System.Linq;

public sealed record MailCapabilityCatalogEntry(
    string capabilityKey,
    string category,
    string surface
);

public static class MailCapabilityCatalog
{
    public static readonly IReadOnlyList<MailCapabilityCatalogEntry> Entries =
    [
        new("transport.connect", "required-baseline", "control-plane"),
        new("transport.authenticate", "required-baseline", "control-plane"),
        new("health", "required-baseline", "control-plane"),
        new("smtp.send", "optional-advanced", "runtime-bridge"),
        new("imap.sync", "optional-advanced", "runtime-bridge"),
        new("imap.folder-sync", "optional-advanced", "runtime-bridge"),
        new("imap.message-sync", "optional-advanced", "runtime-bridge"),
        new("transport.retry", "optional-advanced", "control-plane"),
        new("transport.pool", "optional-advanced", "control-plane"),
    ];

public static IReadOnlyList<MailCapabilityCatalogEntry> GetMailCapabilityCatalog() =>
        Entries;

    public static MailCapabilityCatalogEntry? GetMailCapabilityDescriptor(string capabilityKey) =>
        Entries.FirstOrDefault(entry => entry.capabilityKey == capabilityKey);

}
